daemon/api_systems.go

```diff
@@ -325,7 +325,21 @@ func postSystemActionInstall(c *Command, systemLabel string, req *systemActionRe
 		ensureStateSoon(st)
 		return AsyncResponse(nil, chg.ID())
 	case client.InstallStepFinish:
-		chg, err := devicestateInstallFinish(st, systemLabel, req.OnVolumes)
+		var optional *devicestate.OptionalInstall
+		if req.OptionalInstall != nil {
+			if req.OptionalInstall.All {
+				if len(req.OptionalInstall.Components) > 0 || len(req.OptionalInstall.Snaps) > 0 {
+					return BadRequest("cannot specify both all and individual optional snaps and components to install")
+				}
+			} else {
+				optional = &devicestate.OptionalInstall{
+					Snaps:      req.OptionalInstall.Snaps,
+					Components: req.OptionalInstall.Components,
+				}
+			}
+		}
+
+		chg, err := devicestateInstallFinish(st, systemLabel, req.OnVolumes, optional)
```

Comment:
Is the case of all=true being considered? In that case optional should be set to the available snaps/components.

Reply:
If `all == true`, then we pass `nil` to `devicestate.InstallFinish`. I was trying to avoid having two ways of conveying to that function that we should install everything, but maybe that is too implicit?

Reply:
Right, maybe a bit too implicit, but I think that adding a comment would be enough to clarify.